lib/mp4/transmuxer.js

```diff
@@ -274,24 +340,318 @@ VideoSegmentStream = function(track) {
     // throwing away hundreds of media segment fragments
     boxes = new Uint8Array(moof.byteLength + mdat.byteLength);
 
-    // bump the sequence number for next time
+    // Bump the sequence number for next time
     sequenceNumber++;
 
     boxes.set(moof);
     boxes.set(mdat, moof.byteLength);
 
-    clearDtsInfo(track);
     this.trigger('data', {track: track, boxes: boxes});
 
+    this.resetStreamStage_();
+
+    // Continue with the flush process now
+    this.trigger('done');
+  };
+
+  this.resetStreamStage_ = function() {
+    clearDtsInfo(track);
+
     // reset config and pps because they may differ across segments
     // for instance, when we are rendition switching
     config = undefined;
     pps = undefined;
+  };
 
-    // Continue with the flush process now
-    this.trigger('done');
+  // Search for a candidate Gop for gop-fusion from the gop cache and
+  // return it or return null if no good candidate was found
+  this.getGopForFusion_ = function (nalUnit) {
+    var
+      halfSecond = 45000, // Half-a-second in a 90khz clock
+      allowableOverlap = 9009, // About 3 frames
+      nearestDistance = Infinity,
+      dtsDistance,
+      nearestGopObj,
+      currentGop,
+      currentGopObj,
+      i;
+
+    // Search for the GOP nearest to the beginning of this nal unit
+    for (i = 0; i < this.gopCache.length; i++) {
+      currentGopObj = this.gopCache[i];
+      currentGop = currentGopObj.gop;
+
+      // Reject Gops with different SPS or PPS
+      if (!(track.pps && compareArrays(track.pps[0], currentGopObj.pps[0])) ||
+          !(track.sps && compareArrays(track.sps[0], currentGopObj.sps[0]))) {
+        continue;
+      }
+
+      // Reject Gops that would require a negative baseMediaDecodeTime
+      if (currentGop.dts < track.timelineStartInfo.dts) {
+        continue;
+      }
+
+      // The distance between the end of the gop and the start of the nalUnit
+      dtsDistance = (nalUnit.dts - currentGop.dts) - currentGop.duration;
```

Comment:
`duration` is actually calculated from `dts`. AFAIK, it has always been: https://github.com/videojs/mux.js/blob/master/lib/mp4/transmuxer.js#L215

I actually had a bug early on where I used `pts` instead of `dts` for calculating the `duration` and the results were very, very wrong when given a ts with b-frames.